homeassistant/components/zwave_js/light.py

```diff
@@ -499,61 +542,137 @@ def __init__(
+            return 0
+
+        # Brightness is encoded in the color channels by scaling them lower than 255
+        color_values = [v.value for v in self._get_color_values() if v is not None]
+        max_value = max([v for v in color_values if v is not None])
+        return max_value if max_value is not None else 0
 
     async def async_turn_on(self, **kwargs: Any) -> None:
         """Turn the device on."""
+
         if (
             kwargs.get(ATTR_RGBW_COLOR) is not None
             or kwargs.get(ATTR_COLOR_TEMP) is not None
-            or kwargs.get(ATTR_HS_COLOR) is not None
         ):
+            # RGBW and color temp are not supported in this mode,
+            # delegate to the parent class
             await super().async_turn_on(**kwargs)
             return
 
         transition = kwargs.get(ATTR_TRANSITION)
-        # turn on light to last color if known, otherwise set to white
-        if self._last_color is not None:
-            await self._async_set_colors(
-                {
-                    ColorComponent.RED: self._last_color["red"],
-                    ColorComponent.GREEN: self._last_color["green"],
-                    ColorComponent.BLUE: self._last_color["blue"],
-                },
-                transition,
-            )
-        else:
-            await self._async_set_colors(
-                {
+        brightness = kwargs.get(ATTR_BRIGHTNESS)
+        hs_color = kwargs.get(ATTR_HS_COLOR)
+        new_colors: dict[ColorComponent, int] | None = None
+        scale: float | None = None
+
+        if brightness is None and hs_color is None:
+            # Turned on without specifying brightness or color
+            if self._last_on_color is not None:
+                if self._target_brightness:
+                    # Color is already set, use the binary switch to turn on
+                    await self._async_set_brightness(None, transition)
+                    return
+
+                # Preserve the previous color
+                new_colors = self._last_on_color
+            elif self._supports_color:
+                # Turned on for the first time. Make it white
+                new_colors = {
                     ColorComponent.RED: 255,
                     ColorComponent.GREEN: 255,
                     ColorComponent.BLUE: 255,
-                },
-                transition,
-            )
+                }
+        elif brightness is not None:
+            # If brightness gets set, preserve the color and mix it with the new brightness
+            if self.color_mode == ColorMode.HS:
+                scale = brightness / 255
+            if (
+                self._last_on_color is not None
+                and None not in self._last_on_color.values()
+            ):
+                # Changed brightness from 0 to >0
+                old_brightness = max(self._last_on_color.values())
+                new_scale = brightness / old_brightness
```

Comment:
I had to do this because `scale` is considered `float | None`, even after the definitive assignment here. The suggestion causes an error on the line where `scale` is multiplied.

Maybe there's a different trick I'm not aware of, like TypeScript's non-null assertion?

Reply:
Ok.